lib/ProductOpener/Auth.pm

```diff
@@ -0,0 +1,743 @@
+	param('length', 'logout');
+	init_user($request_ref);
+
+	return $cookie_ref{'return_url'};
+}
+
+sub create_user_in_keycloak ($user_ref, $password) {
+	my $keycloak_users_endpoint = $oidc_options{keycloak_users_endpoint};
+	unless ($keycloak_users_endpoint) {
+		display_error_and_exit('keycloak_users_endpoint not configured', 500);
+	}
+
+	my $token = get_token_using_client_credentials();
+	unless ($token) {
+		display_error_and_exit('Could not get token to manage users with keycloak_users_endpoint', 500);
+	}
+
+	my $api_request_ref = {
+		email => $user_ref->{email},
+		emailVerified => $JSON::PP::true,    # TODO: Keep this for compat with current register endpoint?
+		enabled => $JSON::PP::true,
+		username => $user_ref->{userid},
+		credentials => [
+			{
+				type => 'password',
+				temporary => $JSON::PP::false,
+				value => $password
+			}
+		],
+		attributes => [
+			name => [$user_ref->{name}],
+			locale => [$user_ref->{initial_lc}],
+			country => [$user_ref->{initial_cc}],
+		]
+	};
+	my $json = encode_json($api_request_ref);
+
+	my $create_user_request = HTTP::Request->new(POST => $keycloak_users_endpoint);
+	$create_user_request->header('Content-Type' => 'application/json');
+	$create_user_request->header('Authorization' => $token->{token_type} . ' ' . $token->{access_token});
+	$create_user_request->content($json);
+	my $new_user_response = LWP::UserAgent::Plugin->new->request($create_user_request);
+	unless ($new_user_response->is_success) {
+		display_error_and_exit($new_user_response->content, 500);
+	}
+
+	my $get_user_request = HTTP::Request->new(GET => $new_user_response->header('location'));
+	$get_user_request->header('Content-Type' => 'application/json');
+	$get_user_request->header('Authorization' => $token->{token_type} . ' ' . $token->{access_token});
+	my $get_user_response = LWP::UserAgent::Plugin->new->request($get_user_request);
+	unless ($get_user_response->is_success) {
+		display_error_and_exit($get_user_response->content, 500);
+	}
+
+	my $json_response = $get_user_response->decoded_content(charset => 'UTF-8');
+	my @created_users = decode_json($json_response);
+	return $created_users[0];
+}
+
+sub create_user_in_product_opener ($id_token) {
+	unless ($id_token) {
+		return;
+	}
+
+	my $user_ref = {};
+	$user_ref->{email} = $id_token->{'email'};
+	$user_ref->{userid} = $id_token->{'preferred_username'};
+	$user_ref->{name} = $id_token->{'name'};
+
+	my $user_id = $user_ref->{userid};
+	my $user_file = "$BASE_DIRS{USERS}/" . get_string_id_for_lang("no_language", $user_id) . ".sto";
+	store($user_file, $user_ref);
+
+	# Store email
+	my $emails_ref = retrieve("$BASE_DIRS{USERS}/users_emails.sto");
+	my $email = $user_ref->{email};
+
+	if ((defined $email) and ($email =~ /\@/)) {
+		$emails_ref->{$email} = [$user_id];
+	}
+
+	if (defined $user_ref->{old_email}) {
+		delete $emails_ref->{$user_ref->{old_email}};
+		delete $user_ref->{old_email};
+	}
+
+	store("$BASE_DIRS{USERS}/users_emails.sto", $emails_ref);
```

Comment:
It's a pity to have it repeated in User.pm, can we instead create a sub for that ?

Reply:
Auth.pm now uses a bit different approach now, because I changed the mail handling for users from OIDC.